crates/m3/src/gadgets/mul.rs

```diff
@@ -0,0 +1,431 @@
+	const BIT_LENGTH: usize,
+> {
+	x_in_bits: [Col<B1>; BIT_LENGTH],
+	y_in_bits: [Col<B1>; BIT_LENGTH],
+	out_high_bits: [Col<B1>; BIT_LENGTH],
+	out_low_bits: [Col<B1>; BIT_LENGTH],
+
+	pub x_in: Col<FP>,
+	pub y_in: Col<FP>,
+	pub out_high: Col<FP>,
+	pub out_low: Col<FP>,
+
+	_marker: PhantomData<(FExpBase, UX)>,
+}
+
+impl<
+		FExpBase: TowerField,
+		FP: TowerField,
+		UX: UnsignedMulPrimitives<FP = FP, FExpBase = FExpBase>,
+		const BIT_LENGTH: usize,
+	> Mul<FExpBase, FP, UX, BIT_LENGTH>
+where
+	FExpBase: ExtensionField<FP> + ExtensionField<B1>,
+	B128: ExtensionField<FExpBase> + ExtensionField<FP> + ExtensionField<B1>,
+{
+	// TODO: Figure out if we need to create a separate table for this gadget.
+
+	pub fn new(table: &mut TableBuilder) -> Self {
+		let x_in_bits = table.add_committed_multiple("x_in_bits");
+		let y_in_bits = table.add_committed_multiple("y_in_bits");
+
+		Self::with_inputs(table, x_in_bits, y_in_bits)
+	}
+
+	pub fn with_inputs(
+		table: &mut TableBuilder,
+		x_in_bits: [Col<B1>; BIT_LENGTH],
+		y_in_bits: [Col<B1>; BIT_LENGTH],
+	) -> Self {
+		assert_eq!(FExpBase::TOWER_LEVEL, FP::TOWER_LEVEL + 1);
+		assert_eq!(BIT_LENGTH, 1 << FP::TOWER_LEVEL);
+		assert_eq!(BIT_LENGTH, UX::get_bit_length());
+		// These are currently the only bit lengths I've tested
+		assert!(BIT_LENGTH == 32 || BIT_LENGTH == 64);
+
+		let x_in = table.add_computed("x_in", pack_fp(x_in_bits));
+		let y_in = table.add_computed("y_in", pack_fp(y_in_bits));
+
+		let generator = UX::get_generator().into();
+		let generator_pow_bit_len = UX::get_shifted_generator().into();
+
+		let g_pow_x = table.add_static_exp::<FExpBase>("g^x", &x_in_bits, generator);
+		let g_pow_xy = table.add_dynamic_exp::<FExpBase>("(g^x)^y", &y_in_bits, g_pow_x);
+
+		let out_high_bits = table.add_committed_multiple("out_high");
+		let out_low_bits = table.add_committed_multiple("out_low");
+
+		let out_high = table.add_computed("out_high", pack_fp(out_high_bits));
+		let out_low = table.add_computed("out_low", pack_fp(out_low_bits));
+
+		let g_pow_out_low: Col<FExpBase> =
+			table.add_static_exp("g^(out_low)", &out_low_bits, generator);
+		let g_pow_out_high: Col<FExpBase> = table.add_static_exp(
+			"(g^(2^BIT_LENGTH))^(out_high)",
+			&out_high_bits,
+			generator_pow_bit_len,
+		);
+
+		table.assert_zero("order_non_wrapping", x_in_bits[0] * y_in_bits[0] - out_low_bits[0]);
+		table.assert_zero("exponentiation_equality", g_pow_xy - g_pow_out_low * g_pow_out_high);
+
+		Self {
+			x_in_bits,
+			y_in_bits,
+			out_high_bits,
+			out_low_bits,
+			x_in,
+			y_in,
+			out_high,
+			out_low,
+			_marker: PhantomData,
+		}
+	}
+
+	#[inline]
+	fn populate_internal<P>(
+		&self,
+		index: &mut TableWitnessSegment<P>,
+		x_vals: impl IntoIterator<Item = FP>,
+		y_vals: impl IntoIterator<Item = FP>,
+		fill_input_bits: bool,
+	) -> Result<()>
+	where
+		P: PackedField<Scalar = B128> + PackedExtension<B1> + PackedExtension<FP>,
+	{
+		let mut x_in_bits = self
```

Comment:
Can you use [array_util::try_map](https://docs.rs/array-util/latest/array_util/fn.try_map.html) here instead?